File: LeetCode/0251.Flatten-2d-Vector/Flatten-2d-Vector.py
class Vector2D(object):

    def __init__(self, v):
        """
        :type v: List[List[int]]
        """
        self.stack = []
        for i in range(len(v) - 1, -1, -1):
            self.stack.append(v[i])
        

    def next(self):
        """
        :rtype: int
        """
        self.hasNext()
        return self.stack.pop()[0]
    
    def hasNext(self):
        """
        :rtype: bool
        """
        while self.stack:
            if len(self.stack[-1]) == 1:
                return True
            cur = self.stack.pop()
            for i in range(len(cur) - 1, -1, -1):
                self.stack.append([cur[i]])
        #if self.stack: return True
        return False
        


# Your Vector2D object will be instantiated and called as such:
# obj = Vector2D(v)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: LeetCode/0251.Flatten-2d-Vector/Flatten-2d-Vector.java
class Vector2D {
    int row, col;
    int [][] list;
    public Vector2D(int[][] v) {
        // list = new int [v.length][v[0].length];
        list = v;
        row = 0;
        col = 0;
    }
    
    public int next() {
        update();
        int val = list[row][col];
        col++;
        return val;

    }
    
    public boolean hasNext() {
        update();
        return row != list.length;

    }
    public void update(){
        while(row < list.length && col == list[row].length){
            row++;
            col = 0;
        }
    }
}

/**
 * Your Vector2D object will be instantiated and called as such:
 * Vector2D obj = new Vector2D(v);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */
